esmvalcore/preprocessor/_supplementary_vars.py

```diff
@@ -101,39 +101,72 @@ def add_cell_measure(
-    the cube.
-
     Parameters
     ----------
     cube:
         Iris cube with input data.
     ancillary_cube:
-        Iris cube with ancillary data.
+        Iris cube or AncillaryVariable with ancillary data.
 
     Returns
     -------
     iris.cube.Cube
         Cube with added ancillary variables
     """
-    coord_dims = tuple(range(cube.ndim - len(ancillary_cube.shape), cube.ndim))
-    ancillary_data = ancillary_cube.core_data()
+    try:
+        ancillary_var = iris.coords.AncillaryVariable(
+            ancillary_cube.core_data(),
+            standard_name=ancillary_cube.standard_name,
+            units=ancillary_cube.units,
+            var_name=ancillary_cube.var_name,
+            attributes=ancillary_cube.attributes,
+        )
+    except AttributeError as err:
+        msg = (
+            f"Failed to add {ancillary_cube} to {cube} as ancillary var."
+            "ancillary_cube should be either an iris.cube.Cube or an "
+            "iris.coords.AncillaryVariable object."
+        )
+        raise ValueError(msg) from err
+    data_dims: list[int | None] = []
+    if isinstance(ancillary_cube, iris.coords.AncillaryVariable):
+        start_dim = cube.ndim - len(ancillary_var.shape)
+        data_dims = list(range(start_dim, cube.ndim))
+    else:
+        data_dims = [None] * ancillary_cube.ndim
+        for coord in ancillary_cube.coords():
+            try:
+                for ancillary_dim, cube_dim in zip(
+                    ancillary_cube.coord_dims(coord),
+                    cube.coord_dims(coord),
+                    strict=False,
+                ):
+                    data_dims[ancillary_dim] = cube_dim
+            except iris.exceptions.CoordinateNotFoundError:
+                logger.debug(
+                    "%s from ancillary cube not found in cube coords.", coord
+                )
+        if None in data_dims:
+            none_dims = ", ".join(
+                str(i) for i, d in enumerate(data_dims) if d is None
+            )
+            msg = (
+                f"Failed to add {ancillary_cube} to {cube} as ancillary var."
+                f"No coordinate associated with ancillary cube dimensions"
+                f"{none_dims}"
+            )
+            raise ValueError(msg)
     if ancillary_cube.has_lazy_data():
-        cube_chunks = tuple(cube.lazy_data().chunks[d] for d in coord_dims)
+        cube_chunks = tuple(cube.lazy_data().chunks[d] for d in data_dims)
+        ancillary_data = ancillary_cube.core_data()
         ancillary_data = ancillary_data.rechunk(cube_chunks)
-    ancillary_var = iris.coords.AncillaryVariable(
-        ancillary_data,
-        standard_name=ancillary_cube.standard_name,
-        units=ancillary_cube.units,
-        var_name=ancillary_cube.var_name,
-        attributes=ancillary_cube.attributes,
-    )
-    cube.add_ancillary_variable(ancillary_var, coord_dims)
+        ancillary_var.data = ancillary_data
```

Comment:
```suggestion
        ancillary_var.data = ancillary_cube.lazy_data().rechunk(cube_chunks)
```

Simpler, you can also remove the lines above (cannot do this here somehow...)